### PR TITLE
Security OIDC branch

### PR DESCRIPTION
**Pull request by irinaponomarevahaulmont**
Please review the first draft.

### CODE REVIEW

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+
+Use the *_Plus_* button in the Explorer and select *_Spring Configurations_* -> *_Spring Security Configuration_*.
+
+image::security-oidc/create-security-configuration.png[align=center, width=600]
+
+The following window will appear:
+
+image::security-oidc/create-security-configuration-window.png[align=center, width=700]
+
+Choose the *_OAuth 2.0 / OIDC Authentic ation (stateful, works with Facebook, Google, Github, Keycloak and Okta)_* radiobutton and click *_Next_*.
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:249]: > Authentic ation

typo

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+
+On the second step we fill in the following parameters:
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:251]: You forgot Client ID

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+Setting up authentication in React UI is carried out via *_Plus_* -> *_Frontend_* -> *_Frontend Authentication Scheme_*.
+
+image::security-oidc/frontend-authentication-scheme.png[align=center, width=500]
+
+A wizard will open. At the first step we choose *_OAuth2 flow with Keycloak_* and click *_OK_*.
+
+image::security-oidc/frontend-authentication.png[align=center, width=600]
+
+All communications between the application and the Keycloak server is done via backend.
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:294]: ..., so we don't need to specify any Keycloak connection settings for the frontend.

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+[source, java]
+    @Bean
+    public GrantedAuthoritiesMapper authoritiesMapper() {
+        OAuth2LoginAuthoritiesMapper authoritiesMapper = new OAuth2LoginAuthoritiesMapper();
+        authoritiesMapper.setAuthoritiesClaimName("roles");
+        return authoritiesMapper;
+    }
+    }
+
+ Шеы `"roles"` parameter must match the xref:#token-claim-name[*_Token Claim Name_*] parameter, which we have set in Keycloak while setting up the *_Client scope_*.
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:321]: > Шеы

typo

Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:321]: bad formatting of this paragraph

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+In the Spring Security library each role is a special kind of `authority` with the `ROLE_` prefix. Therefore, any role (e.g. `USER`) received from Keycloak turns into an `authority` after conversion. Its name in our example will be `ROLE_USER`.
+
+This prefix is built into this bean as a default value, as we can see if we jump to source for the `OAuth2LoginAuthoritiesMapper` class:
+
+[source]
+    public OAuth2LoginAuthoritiesMapper(String authoritiesClaimName) {
+        this(authoritiesClaimName, "ROLE_");
+    }
+
+This ensures automatica addition of the prefix to all role names.
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:334]: > automatica

typo

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+
+Click *_Next_*.
+
+Now we need to fill in the following three fields:
+
+* *_Valid redirect URIs_*
+* *_Valid post logout redirect URIs_*
+* *_Web origins_*.
+
+*_Valid redirect URIs_* should contain the mask for the URI where our web client is to be run, e.g. `http://localhost:3000/*`. This value defines, which application will be allowed to use the Keycloak authentication. It should also contain the mask for the backend URI, e.g. `http://localhost:8080/*`.
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:99]: bad formatting of the remaining paragraph after "http://localhost:3000/*"

**Review by alexbudarov — CHANGES_REQUESTED** [review]

**Comment by irinaponomarevahaulmont** [comment]
Requested changes applied, please review.

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+* Identity and access management
+* Social login
+* Two-factor authentication
+* Support of standard protocols
+* Extension support.
+
+[[how-use-keycloak]]
+== How We Use Keycloak
+
+When this type of authentication is set up, all information about the users, including their logins, passwords, other attributes and assigned roles are stored in a separately deployed Keycloak server. The application that we are developing will not contain any information about users, and the application's backend will play the role of the OAuth2 Client. https://docs.spring.io/spring-security/reference/servlet/oauth2/index.html#oauth2-client [See Spring documentation for more details].
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:21]: Bad formatting for the link

File: content/modules/studio/pages/stateful-oidc-authentication.adoc (hunk 1 above)
Comment (alexbudarov) [reply content/modules/studio/pages/stateful-oidc-authentication.adoc]: > Authentic action 

OMG

Authentication!

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+image::security-oidc/create-security-configuration-window.png[align=center, width=700]
+
+Choose the *_OAuth 2.0 / OIDC Authentic action (stateful, works with Facebook, Google, Github, Keycloak and Okta)_* radiobutton and click *_Next_*.
+
+On the second step we fill in the following parameters:
+
+* *_Provider_*: `Keycloak`
+* *_Base URI_*: `http://localhost:8180` (the address through which we access Keycloak on our host machine)
+* *_Realm_*: `tutorials` (the Keycloak realm name)
+* *_Client ID_*: `oidcdemo` (the name of the Amplicode-based application)
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:256]: > the name of the Amplicode-based application


Incorrect. It is the value that we entered previously in KeyCloak.

File: content/modules/studio/pages/stateful-oidc-authentication.adoc [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+Its `"roles"` parameter must match the xref:#token-claim-name[*_Token Claim Name_*] parameter, which we have set in Keycloak while setting up the *_Client scope_*.
+
+The `authorityPrefix` field is responsible for converting the role name into the format required by the Spring Security library.
+
+In the Spring Security library each role is a special kind of `authority` with the `ROLE_` prefix. Therefore, any role (e.g. `USER`) received from Keycloak turns into an `authority` after conversion. Its name in our example will be `ROLE_USER`.
+
+This prefix is built into this bean as a default value, as we can see if we jump to source for the `OAuth2LoginAuthoritiesMapper` class:
+
+[source]
```
Comment (alexbudarov) [inline content/modules/studio/pages/stateful-oidc-authentication.adoc:330]: why no code highlighting in this snippet?

**Review by alexbudarov — CHANGES_REQUESTED** [review]

**Comment by irinaponomarevahaulmont** [comment]
Requested changes applied, please review.

**Pull request merged**